### ISSUE TITLE
[Feature] Terminate on NaN handler

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#201]
What do you think about to add a handler that stops the training when Nan is encountered in the loss output? It is not that complicated to check with a custom handler but as such handlers can be found in other popular frameworks, maybe it worse to add something similar in `ignite` ?

**Comment by alykhantejani** [comment]
Wouldn't this be the same as just doing:

```
trainer.add_event_handler(Events.EXCEPTION_RAISED, lambda x: x.terminate())
```

**Comment by vfdev-5** [comment]
Actually, if loss is nan I'm not sure that the training fails with an exception. For example, mnist.py with huge `lr`
```
ignite/examples# python3 mnist.py --lr=100000
Epoch[1] Iteration[10/938] Loss: nan
Epoch[1] Iteration[20/938] Loss: nan
Epoch[1] Iteration[30/938] Loss: nan
Epoch[1] Iteration[40/938] Loss: nan
Epoch[1] Iteration[50/938] Loss: nan
Epoch[1] Iteration[60/938] Loss: nan
Epoch[1] Iteration[70/938] Loss: nan
Epoch[1] Iteration[80/938] Loss: nan
Epoch[1] Iteration[90/938] Loss: nan
....
Epoch[1] Iteration[910/938] Loss: nan
Epoch[1] Iteration[920/938] Loss: nan
Epoch[1] Iteration[930/938] Loss: nan
Training Results - Epoch: 1  Avg accuracy: 0.10 Avg loss: nan
Validation Results - Epoch: 1  Avg accuracy: 0.10 Avg loss: nan
Epoch[2] Iteration[10/938] Loss: nan
Epoch[2] Iteration[20/938] Loss: nan
Epoch[2] Iteration[30/938] Loss: nan
Epoch[2] Iteration[40/938] Loss: nan
```

**Comment by alykhantejani** [comment]
ah ok, I see what you mean now. 

This could probably be integrated into the `EarlyStoppingHandler` I would imagine

**Comment by vfdev-5** [comment]
@alykhantejani I would say that they have little bit different behaviours. Probably, one can get this with `EarlyStoppingHandler` in a hacky manner, but I think it is more simple to provide built-in handler like
```python
handler = StopOnNanLossHandler()
trainer.add_event_handler(Events.ITERATION_COMPLETED, handler)
```
with 
```
import math
import logging


class StopOnNanLossHandler(object):
    
    def __init__(self):
        self._logger = logging.getLogger(__name__ + "." + self.__class__.__name__)
        self._logger.addHandler(logging.StreamHandler())
        
    def __call__(self, engine):
        loss = engine.state.output        
        if math.isnan(loss):
            self._logger.warn("{}: Loss is Nan. Stop training".format(self.__class__.__name__))
            engine.terminate()


```

**Comment by vfdev-5** [comment]
As discussed in slack, probably a correct behaviour in a situation with NaN/Inf loss is too stop the training:
> NaN+anything=NaN, so you’d just continue getting NaN…for it to be useful you need a mechanism to roll-back to a previous value which was not NaN

**Comment by vfdev-5** [comment]
A solution can be proposed using [anomaly detection](https://pytorch.org/docs/stable/autograd.html#anomaly-detection) introduced in pytorch 0.4.1
